Realm/Realm.SourceGenerator/InfoClasses.cs

```diff
@@ -240,22 +242,101 @@ public bool IsSupportedRequiredType()
             return _requiredTypes.Contains(ScalarType);
         }
 
-        public bool HasCorrectNullabilityAnnotation()
+        public bool HasCorrectNullabilityAnnotation(bool ignoreObjectsNullability)
         {
             if (NullableAnnotation == NullableAnnotation.Annotated &&
                 (IsCollection || ScalarType == ScalarType.RealmValue))
             {
                 return false;
             }
 
-            if (!IsNullable &&
-                (ScalarType == ScalarType.Object))
+            if (!ignoreObjectsNullability)
             {
-                return false;
+                if (!IsNullable && ScalarType == ScalarType.Object)
+                {
+                    return false;
+                }
+
+                if (IsCollection && InternalType.ScalarType == ScalarType.Object)
+                {
+                    if (IsDictionary && InternalType.NullableAnnotation == NullableAnnotation.NotAnnotated)
+                    {
+                        return false;
+                    }
+
+                    if ((IsList || IsSet || IsBacklink) && InternalType.NullableAnnotation == NullableAnnotation.Annotated)
+                    {
+                        return false;
+                    }
+                }
             }
 
             return true;
         }
+
+        /** This method returns the type with the type string with correct nullability annotation wether they were enabled or not in the original file,
+         * together with the same annotation for the internal type (for collections). In some cases we were just getting the internal type string on his own, and this would be wrong for some collections.
+         * Simplified example from a generated unmanaged accessor:
+         * <code>
+         * public IList≪AllTypesClass≫ ObjectCollectionProperty
+         * {
+         *      get => GetListValue≪AllTypesClass≫("ObjectCollectionProperty")
+         * }
+         * </code>
+         * If we would take the nullability of AllTypesClass on his own, this would be nullable. This can't be nullable though, because it can't be nullable in a list.
+         */
+        public (string CompleteType, string InternalType) GetCorrectlyAnnotatedTypeName(bool isRequired)
+        {
+            var internalTypeString = InternalType?.CompleteFullyQualifiedString;
+            var nullableInternalTypeString = $"{internalTypeString}?";
+
+            if (!isRequired)
+            {
+                if (NullableAnnotation == NullableAnnotation.None && (ScalarType == ScalarType.Data || ScalarType == ScalarType.String))
+                {
+                    return (CompleteFullyQualifiedString + "?", null);
+                }
+
+                if (IsCollection && InternalType.NullableAnnotation == NullableAnnotation.None
```

Comment:
This warning is correct - if we're using `InternalType?.CompleteFullyQualifiedString`, it means we're expecting that `InternalType` may be null. If that's the case, we should guard here as well.

Reply:
Hm... I guess it's not smart enough to figure that out :/ Would have been nice if we have nullable reference types, but 🤷‍♂️